docs/website/docs/extensions/index.md

```diff
@@ -0,0 +1,442 @@
+# Extension mechanisms
+
+**NOTE**: much of this describes provisions for extension within IREE but until
+the core of the system has settled little work will be done to fully flesh-out
+and document them in detail. A large majority of things that would make someone
+want to extend IREE can instead be accomplished much easier and performantly
+using native MLIR dialects that are then processed by the IREE compiler.
+
+IREE has a compiler and runtime separation, a multi-layered architecture, and
+split between execution of "host code" that schedules compute-heavy work and
+[SPMD](https://en.wikipedia.org/wiki/SPMD) "device code" that performs the bulk
+of compute operations. Each axis has a different set of extension mechanisms
+that can be used independently or combined.
+
+The primary goal when extending any framework should first be to avoid extending
+it at all. There is no mechanism that is free - whether in terms of engineering
```

Comment:
This appears as a pretty big wall of text on the rendered site:

![image](https://user-images.githubusercontent.com/4010439/139780729-f2756e2a-8b7c-40de-9a3e-06133715911f.png)

Diagrams, images, bullet points, emojis/icons, admonitions (like "tip" or "warning") or more paragraph splits could help break it up and make it easier to scan. I haven't done a full read through yet, so I may come up with more concrete suggestions later.

Reply:
Some more thoughts after reading through:

The words here are all important for providing context and high level guidance on pros/cons of various approaches. If someone is seriously considering an extension, they should take the time to read through and evaluate each option. That being said, breaking up into bullet points or sections ("why extend?", "performance fundamentals", etc.) would still help with scanning.

I'm fine with landing as-is though.

Reply:
will split up into sections like you recommend below